[dtensor] add more documentations

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133113
* __->__ #133306
* #133305

This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

cc @XilunWu @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

Differential Revision: [D61458681](https://our.internmc.facebook.com/intern/diff/D61458681)

wanchaol committed:
[dtensor] add more documentations

This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add more documentations"


This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add more documentations"


This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add more documentations"


This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add more documentations"


This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add more documentations"


This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/_dispatch.py:
```diff
@@ -71,6 +71,13 @@ class OpDispatcher:
+    is designed to minimize the CPU overhead by using the tricks of proper unflattening, faster
+    pytree if needed, and leveraging various caching mechanisms implemented in the sharding
+    propagation and redistribute modules. The CPU overhead is critical to eager mode performance,
+    one need to carefully measure the CPU overhead when making significant changes to the
+    OpDispatcher and ShardingPropagator.
```
is there any suggestion on CPU overhead measurement? like, by running profiling in model testing?

wanchaol replied:
yeah running profiling with some benchmark scripts to measure runtime overhead

XilunWu commented on torch/distributed/_tensor/api.py:
```diff
@@ -216,13 +216,30 @@ def backward(ctx, grad_output: "DTensor"):  # type: ignore[override]
-class DTensor(torch.Tensor):  # pyre-ignore[13]: pyre is bad at __new__
+class DTensor(torch.Tensor):
+    """
+    ``DTensor`` (Distributed Tensor) is a subclass of ``torch.Tensor`` that provides single-device like
```
"single-device" -> "a single-device"

XilunWu commented on torch/distributed/_tensor/api.py:
```diff
@@ -216,13 +216,30 @@ def backward(ctx, grad_output: "DTensor"):  # type: ignore[override]
+    * :class:`Replicate`: Tensor replicated on the devices of the ``DeviceMesh`` dimension
+    * :class:`Partial`: Tensor is pending reduction on the devices of the ``DeviceMesh`` dimension
+
+    When calling PyTorch operators, ``DTensor`` overrides the PyTorch operators to perform sharded computation or issue
```
"or" -> "and"; "issue" -> "issues"

XilunWu commented on torch/distributed/_tensor/api.py:
```diff
@@ -478,7 +510,11 @@ def redistribute(
+            formula of the redistribute operation.
+
+        .. note:: ``redistribute`` currently only supports redistributing DTensor on the same DeviceMesh,
+            Please file a issue if you need to redistribute DTensor to different DeviceMesh.
```
"a issue" -> "an issue"

XilunWu commented on torch/distributed/_tensor/api.py:
```diff
@@ -593,16 +629,19 @@ def distribute_tensor(
-    to the ``placements`` specified. The rank of ``device_mesh`` and ``placements`` must be
-    the same. If you want to construct a DTensor in the middle of the Autograd computation,
-    please use ``DTensor.from_local`` instead.
+    Distribute a leaf torch.Tensor (i.e. nn.Parameter/buffers) to the ``device_mesh`` according
```
"torch.Tensor" -> "\`\`torch.Tensor\`\`"

XilunWu reviewed (approved):
LGTM. Some typos.

tianyu-l commented on torch/distributed/_tensor/api.py:
```diff
@@ -463,13 +481,27 @@ def redistribute(
             device_mesh (:class:`DeviceMesh`, optional): DeviceMesh to place the
-                DTensor, if not specified, must be called under a DeviceMesh
-                context manager, default: None
+                DTensor, if not specified, it would use the current DTensor's DeviceMesh,
```
```suggestion
                DTensor. If not specified, it would use the current DTensor's DeviceMesh.
```

tianyu-l commented on torch/distributed/_tensor/api.py:
```diff
@@ -478,7 +510,11 @@ def redistribute(
             A :class:`DTensor` object
 
-        .. note:: ``redistribute`` is differentiable.
+        .. note:: ``redistribute`` is differentiable, user do not need to worry about the backward
```
```suggestion
        .. note:: ``redistribute`` is differentiable, which means user do not need to worry about the backward
```

wz337 commented on torch/distributed/_tensor/api.py:
```diff
@@ -734,12 +773,13 @@ def distribute_module(
-        specifying the ``input_fn`` and ``output_fn``. (i.e. convert the input to
-        :class:`DTensor`, convert the output back to torch.Tensor)
+    specifying the ``input_fn`` and ``output_fn``. (i.e. convert the input to
+    :class:`DTensor`, convert the output back to torch.Tensor)
```
nit: torch.Tensor -> ``torch.Tensor``

wz337 commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -574,10 +575,10 @@ class Partial(Placement):
-        to produce Replicated/Sharded DTensor. Only element-wise reduction operations
-        are supportd, including: "sum", "avg", "prod", "max", "min", default: "sum".
+            to produce Replicated/Sharded DTensor. Only element-wise reduction operations
+            are supportd, including: "sum", "avg", "prod", "max", "min", default: "sum".
```
nit: supportd -> supported

wanchaol committed:
Update on "[dtensor] add more documentations"


This PR adds more documentations to the DTensor APIs, to prepare for the
module be public

cc XilunWu H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D61458681).
